ncm-nfs/src/main/perl/nfs.pm

```diff
@@ -2,305 +2,468 @@
-
-
-  #########################################
-  # Process exports to build /etc/exports #
-  #########################################
-      
-  if ($nfs_config->{'exports'}) {
-  
-    # Pull out the parameters from the configuration. 
-    my $exports = $nfs_config->{'exports'};
-    
-    # Loop through all of the servers creating a line for each.
-    for my $entry (@{$exports}) {
-      my $path = $entry->{'path'};
-      my $hosts = $entry->{'hosts'};
-      
-      my @entries;
-      for my $host_e (keys(%{$hosts}) ) {
-        my $host = unescape($host_e);
-        my $export_opts = $hosts->{$host_e};
-        # add option 'sync if neither 'sync', nor 'async' has been specified
-        if ( $export_opts !~ m/sync/ ) {
-          if ( length($export_opts) > 0 ) {
-            $export_opts .= ',';
-          }
-          $export_opts .= 'sync';
+
+use File::Path qw(mkpath);
+
+use CAF::Object qw(SUCCESS);
+use CAF::FileWriter;
+use CAF::FileReader;
+use CAF::Service;
+
+use EDG::WP4::CCM::Element qw(unescape);
+
+use Readonly;
+
+our $NoActionSupported = 1;
+
+# The absolute names of configuration files.
+Readonly my $EXPORTS => "/etc/exports";
+Readonly my $FSTAB => "/etc/fstab";
+
+Readonly my $EXPORTS_DEFAULT_SYNC => 'sync';
+
+Readonly my $FSTAB_ACTION => 'action';
+
+Readonly my $FSTAB_DEVICE => 'device';
+Readonly my $FSTAB_MOUNTPOINT => 'mountpoint';
+Readonly my $FSTAB_FSTYPE => 'fstype';
+Readonly my $FSTAB_OPTIONS => 'options';
+Readonly my $FSTAB_FREQ => 'freq';
+Readonly my $FSTAB_PASSNO => 'passno';
+
+Readonly::Array my @FSTAB_LINE => (
+    $FSTAB_DEVICE, $FSTAB_MOUNTPOINT, $FSTAB_FSTYPE,
+    $FSTAB_OPTIONS, $FSTAB_FREQ, $FSTAB_PASSNO,
+);
+
+Readonly::Hash my %FSTAB_DEFAULT => {
+    $FSTAB_OPTIONS => 'defaults',
+    $FSTAB_FREQ => 0,
+    $FSTAB_PASSNO => 0,
+};
+
+Readonly my $FSTAB_ACTION_NONE => 'none';
+Readonly my $FSTAB_ACTION_MOUNT => 'mount';
+Readonly my $FSTAB_ACTION_REMOUNT => 'remount';
+Readonly my $FSTAB_ACTION_UMOUNT_MOUNT => 'umount/mount';
+
+Readonly::Array my @CMD_UMOUNT_LAZY => qw(umount -l);
+Readonly::Array my @CMD_MOUNT => qw(mount);
+Readonly::Array my @CMD_REMOUNT => qw(mount -o remount);
+
+Readonly::Hash my %CLEANUP_DISPATCH => {
+    move => \&move,
+    rmtree => \&rmtree,
+    unlink => sub { return unlink(shift); },
+};
+
+
+
+# Compares two fstab entries C<new> and C<old> for equality,
+# and return action to be taken.
+# If old does not exist, mount.
+# If equal, do nothing.
+# If the entries differ in the devices or mountpoint, do unmount/mount.
+# Otherwise, remount.
+# This is a function, not a method.
+sub mount_action_new_old
+{
+    my ($new, $old) = @_;
+
+    return $FSTAB_ACTION_MOUNT if (! $old);
+
+    # Can numeric 'A != B' be different than 'A ne B' ?
+    my $ne = sub {
+        my $attr = shift;
+        return $new->{$attr} ne $old->{$attr};
+    };
+
+    return $FSTAB_ACTION_UMOUNT_MOUNT if (&$ne($FSTAB_DEVICE) || &$ne($FSTAB_MOUNTPOINT));
```

Comment:
what do you mean with SHOUTING?
i'll try to clean it up again (wasn't that happy with the result myself)